caikit_tgis_backend/managed_tgis_subprocess.py

```diff
@@ -0,0 +1,371 @@
+        self._mutex = Lock()
+        self._tgis_state: _TGISState = _TGISState.STOPPED
+        self._tgis_proc = None
+        self._tgis_client = None
+        self._wrapped_client = None
+        self._bootup_thread = None
+        self._bootup_exc = None
+
+    def __del__(self):
+        self.terminate()
+
+    def is_ready(self) -> bool:
+        """Return true if a request can be propagated to the subprocess"""
+        return self._tgis_state == _TGISState.READY and self._tgis_client
+
+    def get_client(self):
+        """Creates a gRPC client to the subprocess
+
+        Returns None if the subprocess is not ready and doesn't have an internal client
+        """
+        if self._wrapped_client:
+            return self._wrapped_client
+
+        if self._tgis_client:
+            # create a wrapped gRPC client stub that defers to self._tgis_client
+            self._wrapped_client = GenerationServiceStubWrapper(self)
+            return self._wrapped_client
+
+        log.warning("<MTS7717800W>", "get_client called with _tgis_client set to None")
+        return None
+
+    def launch(self):
+        """Launch the subprocess or restart it if it exists"""
+        with self._mutex:
+            self._launch()
+
+    def wait_until_ready(self, timeout=None) -> bool:
+        start_t = time.time()
+
+        # wait for the boot monitoring thread if we have one
+        if self._bootup_thread:
+            self._bootup_thread.join(timeout)
+
+        while True:
+            log.debug("wait_until_ready saw state [%s]", self._tgis_state)
+            if self._tgis_state == _TGISState.READY:
+                return True
+
+            if self._tgis_state == _TGISState.STOPPED:
+                # raise the bootup exception if we captured one
+                if self._bootup_exc:
+                    raise self._bootup_exc
+                return False
+
+            if time.time() - start_t >= timeout:
+                return False
+
+            time.sleep(self._bootup_poll_delay)
+
+    def terminate(self):
+        """Terminate the subprocess, wait for it to exit"""
+        with self._mutex:
+            self._ensure_terminated()
+
+    def _ensure_terminated(self):
+        """Terminate the subprocess if it exists
+
+        _mutex should be locked before calling this
+        """
+        if self._tgis_state != _TGISState.STOPPED:
+            if self._tgis_proc:
+                self._tgis_proc.terminate()
+                self._tgis_proc = None
+                # disown the bootup thread (but keep the exception if one was captured)
+                self._bootup_thread = None
+            log.debug("_ensure_terminated setting the state to STOPPED")
+            self._tgis_state = _TGISState.STOPPED
+
+    def _launch(self):
+        """Launch the subprocess or restart if it exists
+
+        _mutex should be locked before calling
+        """
+        self._ensure_terminated()
+
+        # Launch TGIS
+        launch_cmd = " ".join(
```

Comment:
NIT: I'm wondering if any of this can be stored on `self` to avoid some of the extra instance variables (pesky `pylint`!)

Reply:
Hmm, of course some of it could be. When I run `python -m tox -e lint` I don't see these lines being picked out. What's the lint rule you are seeing? Is there a change that should be made to `.pylintrc` to check for it?

Reply:
I was looking at

```py
# pylint: disable=too-many-instance-attributes
class ManagedTGISSubprocess:
```

Reply:
Heh, yeah, sorry that was just poorly written. You're right, I was wondering if we could reduce the count by combining them into pre-built formatted strings for the commands. _very_ low priority!